Improve AsyncAPI tool learn docs (#4444)
ballerina-platform/ballerina-dev-website: learn/guides/asyncapi-tool.md

  - /learn/ballerina-asyncapi-support
  - /learn/ballerina-asyncapi-support/
  - /learn/asyncapi-tool
---

> Ballerina AsyncAPI tool will make it easy for you to start the development of an event API documented in an AsyncAPI contract in Ballerina by generating a Ballerina service and listener skeletons.

Reviewer: ```suggestion
The Ballerina AsyncAPI tool will make it easy for you to start the development of an event API documented in an AsyncAPI contract in Ballerina by generating a Ballerina service and listener skeletons.
```

> **Prerequisite:** Install the latest [Ballerina Swan Lake distribution](https://ballerina.io/downloads/).

## Generate Ballerina sources from AsyncAPI contracts

> If you prefer the design-first approach, use an existing or your AsyncAPI definition to generate Ballerina sources using the AsyncAPI CLI command below.

Reviewer: ```suggestion
If you prefer the design-first approach, use an existing AsyncAPI definition to generate Ballerina sources using the AsyncAPI CLI command below.
```

```bash
bal asyncapi -i <asyncapi-contract>
```

> The generated sources can be used as a code template to start the service implementation. For example,

Reviewer: ```suggestion
The generated sources can be used as code templates to start the service implementation. For example,
```

Author: Thanks a lot for the suggestions, Resolved all

```bash
bal asyncapi -i hello.yaml
```

> This will generate a Ballerina source (i.e., the four Ballerina files below) from the given AsyncAPI definition file.

Reviewer 2: ```suggestion
This generates a Ballerina source (i.e., the four Ballerina files below) from the given AsyncAPI definition file.
```

              title: When the event was dispatched
              type: string
            type:
              title: The specific name of the event
              type: string
            text:
              title: The message content
              type: string
          required:
            - type
            - event_ts
          title: "The actual event, an object, that happened"
          type: object
      required:
        - event
      title: Standard event wrapper for the Events API
      type: object
channels:
  app:
    subscribe:
      message:
        oneOf:
          - x-ballerina-event-type: "app_mention"
            externalDocs:
              description: Event documentation for app_mention
            payload:
              $ref: "#/components/schemas/GenericEventWrapper"
            summary: Subscribe to only the message events that mention your app or bot
            tags:
              - name: allows_user_tokens
              - name: app_event
            x-scopes-required: []
            x-tokens-allowed:
              - user
```

There are custom tags in this YAML starting with `x-ballerina`. Those tags are being used for the purposes below.

> - `x-ballerina-event-identifier` - When the listener receives an event from the event source (Slack is the event source in this scenario), there should be a way to identify the event type. This includes two parts, `type` and `path`. Type can be either `header` or `body`. In other words, type of the event can be included in the payload either as a header or as an attribute in the body.

Reviewer 2: ```suggestion
- `x-ballerina-event-identifier` - when the listener receives an event from the event source (Slack is the event source in this scenario), there should be a way to identify the event type. This includes two parts, `type` and `path`. Type can be either `header` or `body`. In other words, the type of the event can be included in the payload either as a header or as an attribute in the body.
```

- `x-ballerina-event-type` - This should be there in every event inside the channel. This is the name of the event or the value of the above-mentioned attribute for a specific event.

### Generate Ballerina sources

Execute the command below on the above file. This will generate four files as mentioned above.

```bash
bal asyncapi -i slack.yaml
```

### Modify the generated sources

The generated files can be modified according to the custom requirements. When modifying the generated code segments, it will be easier to consider the facts below.

- All the incoming requests will be coming to the resource function in the `dispatcher_service.bal` file. Hence, if there is a necessity to add an authentication logic for the incoming calls, that logic can be included there before processing the incoming HTTP request.
- If more information is needed when initializing the listener like secrets, endpoint URLs, tokens, refresh tokens, etc., update the `init` function in the `listener.bal` file.

### Execute the generated sources

In order to use the generated Ballerina sources, please follow the below steps,

> - Create an empty folder and copy all the generated files there.

Reviewer 2: ```suggestion
1. Create an empty folder and copy all the generated files to it.
```

> - Execute the below command in the root of the folder

Reviewer 2: ```suggestion
2. Execute the command below from the root of the folder.
```

  ```bash
  bal init
  ```
  this will generate a Ballerina.toml file in the root of the folder.

> - Create a new ballerina file inside the folder (Ex: slack_service.bal ) and paste this piece of code

Reviewer 2: ```suggestion
3. Create a new Ballerina file inside the folder (e.g., `slack_service.bal`) and paste the code below to it.
```

>   ```

Reviewer 2: ```suggestion
  ```ballerina
```

  listener Listener webhookListener = new(8090);

  service AppService on webhookListener {
    remote function onAppMention(GenericEventWrapper event) returns error? {

>         //implement the logic to use the received `event` here.

Reviewer 2: ```suggestion
        //Implement the logic to use the received `event` here.
```

    }
  }
  ```

> - Run the folloewing command in the root of the folder to execute this service.

Reviewer 2: ```suggestion
4. Execute the command below from the root of the folder to execute the generated service.
```
Should it be the generated source or generated service?

Author: Thanks a lot for the above suggestions.
I think, this should be named as service, in the 3rd step we are adding a service.

- `x-ballerina-event-type` - this should be there in every event inside the channel. This is the name of the event or the value of the above-mentioned attribute for a specific event.

### Generate Ballerina sources

Execute the command below on the above file. This will generate four files as mentioned above.

```bash
bal asyncapi -i slack.yaml
```

### Modify the generated sources

The generated files can be modified according to the custom requirements. When modifying the generated code segments, it will be easier to consider the facts below.

- All the incoming requests will be coming to the resource function in the `dispatcher_service.bal` file. Hence, if there is a necessity to add an authentication logic for the incoming calls, that logic can be included there before processing the incoming HTTP request.
- If more information is needed when initializing the listener like secrets, endpoint URLs, tokens, refresh tokens, etc., update the `init` function in the `listener.bal` file.

### Execute the generated sources

Follow the steps below to use the generated Ballerina sources.

> 1\. Create an empty folder and copy all the generated files to it.

Reviewer 2: ```suggestion
1. Create an empty folder and copy all the generated files to it.
```

> 2\. Execute the command below from the root of the folder. This generates a `Ballerina.toml` file in the root of the folder.

Reviewer 2: ```suggestion
2. Execute the command below from the root of the folder. This generates a `Ballerina.toml` file in the root of the folder.
```

```bash
bal init
```

> 3\. Create a new Ballerina file inside the folder (e.g., `slack_service.bal`) and paste the code below to it.

Reviewer 2: ```suggestion
3. Create a new Ballerina file inside the folder (e.g., `slack_service.bal`) and paste the code below to it.
```

```ballerina
listener Listener webhookListener = new(8090);

service AppService on webhookListener {
  remote function onAppMention(GenericEventWrapper event) returns error? {
      //Implement the logic to use the received `event` here.
  }
}
```

> 4\. Execute the command below from the root of the folder to execute this service.

Reviewer 2: ```suggestion
4. Execute the command below from the root of the folder to execute this service.
```

Author: These \ were added to prevent auto numbering.
